src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs

```diff
@@ -0,0 +1,829 @@
+            foreach (var projectBlock in solutionFile.ProjectBlocks)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                string absoluteProjectPath;
+                if (TryGetAbsoluteProjectPath(projectBlock.ProjectPath, solutionFolder, reportMode, out absoluteProjectPath))
+                {
+                    IProjectFileLoader loader;
+                    if (TryGetLoaderFromProjectPath(absoluteProjectPath, reportMode, out loader))
+                    { 
+                        // projects get added to 'loadedProjects' as side-effect
+                        // never prefer metadata when loading solution, all projects get loaded if they can.
+                        var tmp = await GetOrLoadProjectAsync(absoluteProjectPath, loader, preferMetadata: false, loadedProjects: loadedProjects, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    }
+                }
+            }
+#endif
+
+            // construct workspace from loaded project infos
+            return SolutionInfo.Create(SolutionId.CreateNewId(debugName: absoluteSolutionPath), version, absoluteSolutionPath, loadedProjects.Values);
+        }
+
+        internal string GetAbsoluteSolutionPath(string path, string baseDirectory)
+        {
+            string absolutePath;
+
+            try
+            {
+                absolutePath = GetAbsolutePath(path, baseDirectory);
+            }
+            catch (Exception)
+            {
+                throw new InvalidOperationException(string.Format(WorkspacesResources.InvalidSolutionFilePath, path));
+            }
+
+            if (!File.Exists(absolutePath))
+            {
+                throw new FileNotFoundException(string.Format(WorkspacesResources.SolutionFileNotFound, absolutePath));
+            }
+
+            return absolutePath;
+        }
+
+        /// <summary>
+        /// Loads the <see cref="ProjectInfo"/> for the specified project. 
+        /// Referenced projects are converted to metadata references unless those projects are already loaded.
+        /// </summary>
+        public async Task<ProjectInfo> LoadProjectAsync(string projectFilePath, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var loadedProjects = new Dictionary<ProjectId, ProjectInfo>();
+            var pid = await LoadProjectInfoAsync(projectFilePath, loadedProjects, preferMetadata: this.LoadMetadataForReferencedProjects, cancellationToken: cancellationToken).ConfigureAwait(false);
+            return loadedProjects[pid];
+        }
+
+        /// <summary>
+        /// Loads the <see cref="ProjectInfo"/>'s from the specified project file and all referenced projects.
+        /// </summary>
+        public Task<ProjectId> LoadProjectsAsync(string projectFilePath, Dictionary<ProjectId, ProjectInfo> loadedProjects, CancellationToken cancellationToken = default(CancellationToken))
+        {
```

Comment:
This was originally an internal API, but needs to be public now.  It basically needs to return all the ProjectInfo's for all referenced projects. But it also needs to return the ProjectId of the one associated with the project file you specified.  So it cannot just use SolutionInfo.  I don't have anything better yet.  I'm open to suggestions.